habitat-lab/habitat/tasks/rearrange/multi_task/pddl_defined_predicates.py

```diff
@@ -0,0 +1,384 @@
+        # an object is inside of a receptacle.
+        return is_inside(obj, at_entity, sim_info)
+    elif sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.GOAL_ENTITY.value
+    ):
+        targ_idx = cast(
+            int,
+            sim_info.search_for_entity(at_entity),
+        )
+        idxs, pos_targs = sim_info.sim.get_targets()
+        targ_pos = pos_targs[list(idxs).index(targ_idx)]
+
+        dist = float(np.linalg.norm(entity_pos - targ_pos))
+        return dist < dist_thresh
+    elif sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.STATIC_RECEPTACLE_ENTITY.value
+    ):
+        # TODO: Fix this logic to be using
+        # habitat/sims/habitat_simulator/sim_utilities.py
+        recep = cast(mn.Range3D, sim_info.search_for_entity(at_entity))
+        return recep.contains(entity_pos)
+    else:
+        raise ValueError(
+            f"Got unexpected combination of {obj} and {at_entity}"
+        )
+
+
+def set_object_at(
+    obj: PddlEntity,
+    at_entity: PddlEntity,
+    sim_info: PddlSimInfo,
+    recep_place_shrink_factor: float = 0.8,
+) -> None:
+    """
+    Sets a movable PDDL entity to match the transformation of a desired
+    `at_entity` which can be a receptacle or goal.
+
+    :param recep_place_shrink_factor: How much to shrink the size of the
+        receptacle by when placing the entity on a receptacle.
+    """
+
+    sim = sim_info.sim
+
+    # The source object must be movable.
+    if not sim_info.check_type_matches(
+        obj, SimulatorObjectType.MOVABLE_ENTITY.value
+    ):
+        raise ValueError(f"Got unexpected obj {obj}")
+
+    elif sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.GOAL_ENTITY.value
+    ):
+        targ_idx = cast(
+            int,
+            sim_info.search_for_entity(at_entity),
+        )
+        all_targ_idxs, pos_targs = sim.get_targets()
+        targ_pos = pos_targs[list(all_targ_idxs).index(targ_idx)]
+        set_T = mn.Matrix4.translation(targ_pos)
+    elif sim_info.check_type_matches(
+        at_entity, SimulatorObjectType.STATIC_RECEPTACLE_ENTITY.value
+    ):
+        # Place object on top of receptacle.
+        recep = cast(mn.Range3D, sim_info.search_for_entity(at_entity))
+
+        # Divide by 2 because the `from_center` creates from the half size.
+        shrunk_recep = mn.Range3D.from_center(
+            recep.center(),
+            (recep.size() / 2.0) * recep_place_shrink_factor,
+        )
+        pos = np.random.uniform(shrunk_recep.min, shrunk_recep.max)
+        set_T = mn.Matrix4.translation(pos)
+    else:
+        raise ValueError(f"Got unexpected at_entity {at_entity}")
+
+    obj_idx = cast(int, sim_info.search_for_entity(obj))
+    abs_obj_id = sim.scene_obj_ids[obj_idx]
+
+    # Get the object id corresponding to this name
+    rom = sim.get_rigid_object_manager()
+    set_obj = rom.get_object_by_id(abs_obj_id)
+    set_obj.transformation = set_T
```

Comment:
would be good to actually have a function like set_object(set_T) that does all of this to avoid this boilerplate code

Reply:
Yeah I agree, but the new sim utilities will also handle this in the future.